src/Compilers/Test/Core/Assert/AssertEx.cs

```diff
@@ -594,37 +616,117 @@ public static void AssertResultsEqual(string result1, string result2)
+            actualString = NormalizeWhitespace(actualString, out var pooled2);
+
+            if (!actualString.Contains(expectedSubString, StringComparison.Ordinal))
+            {
+                // Call through to actual API to get same error message experience
+                Assert.Contains(expectedSubString.ToString(), actualString.ToString(), StringComparison.Ordinal);
+            }
+
+            if (pooled1 != null)
+            {
+                ArrayPool<char>.Shared.Return(pooled1);
+            }
+            if (pooled2 != null)
+            {
+                ArrayPool<char>.Shared.Return(pooled2);
+            }
         }
 
         public static void AssertStartsWithToleratingWhitespaceDifferences(string expectedSubString, string actualString)
         {
-            expectedSubString = NormalizeWhitespace(expectedSubString);
-            actualString = NormalizeWhitespace(actualString);
-            Assert.StartsWith(expectedSubString, actualString, StringComparison.Ordinal);
+            AssertStartsWithToleratingWhitespaceDifferences(expectedSubString.AsSpan(), actualString.AsSpan());
+        }
+
+        public static void AssertStartsWithToleratingWhitespaceDifferences(ReadOnlySpan<char> expectedSubString, ReadOnlySpan<char> actualString)
+        {
+            expectedSubString = NormalizeWhitespace(expectedSubString, out var pooled1);
+            actualString = NormalizeWhitespace(actualString, out var pooled2);
+
+            if (!actualString.StartsWith(expectedSubString, StringComparison.Ordinal))
+            {
+                // Call through to actual API to get same error message experience
+                Assert.StartsWith(expectedSubString.ToString(), actualString.ToString(), StringComparison.Ordinal);
+            }
+
+            if (pooled1 != null)
+            {
+                ArrayPool<char>.Shared.Return(pooled1);
+            }
+            if (pooled2 != null)
+            {
+                ArrayPool<char>.Shared.Return(pooled2);
+            }
+        }
+
+        internal static ReadOnlySpan<char> NormalizeWhitespace(ReadOnlySpan<char> input, out char[] pooledAlloc)
+        {
+            return NormalizeWhitespaceROM(input, out pooledAlloc).Span;
         }
 
-        internal static string NormalizeWhitespace(string input)
+        internal static ReadOnlyMemory<char> NormalizeWhitespaceROM(ReadOnlySpan<char> input, out char[] pooledAlloc)
         {
-            var output = new StringBuilder();
-            var inputLines = input.Split('\n', '\r');
-            foreach (var line in inputLines)
+            // Setup allocation (start with none)
+            pooledAlloc = null;
+            int allocSize = 0;
+
+            // Loop over each line section until we run out of lines
+            int prev = 0;
+            int next = input.IndexOfAny("\r\n".AsSpan());
+            if (next < 0) next = input.Length;
+            while (true)
```

Comment:
Reading through this I'm wondering if it should be broken up a bit. Basically should we have one method that iterates through a `ReadOnlySpan<char>` and gives us back `ReadOnlyMemory<char>` for the lines within it?

Reply:
I'm not exactly sure what you mean by this?